src/memory.jl

```diff
@@ -1,203 +1,318 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct MemoryDriverState <: AbstractMaterialState
-    time :: Float64 = zero(Float64)
-    strain :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module MemoryModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericMemory, GenericMemoryDriverState, GenericMemoryParameterState, GenericMemoryVariableState
+
+# specialization for Float64
+export Memory, MemoryDriverState, MemoryParameterState, MemoryVariableState
+
+@with_kw mutable struct GenericMemoryDriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for Memory material.
```

Comment:
Strain memory material. Strain memory is used to be able to model strain amplitude-dependent isotropic hardening. In practice, the transition from tensile test curve to cyclic behavior can be captured with this model.